Automatic documentation for experiment and analysis

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
This PR adds auto docstring generation decorator for experiment and analysis classes.

close #92

### Details and comments
The new package `qiskit_experiments.autodocs` is added.

Class implementation has changed. Look this comment.
https://github.com/Qiskit/qiskit-experiments/pull/116#issuecomment-880540352

-- old --
The docstring generator is implemented based on [facade pattern](https://en.wikipedia.org/wiki/Facade_pattern) which is often used for auto generation of, for example, webpage. This includes actual docstring writer class and facade class to define the style of docstring. By adding new facade class, you can define own docstring format for your subset of experiments.

Now base classes have several class attributes `__doc_*__` corresponding to a section of documentation. A developer needs to fill these attributes to generate docstring sections. The allocation of sections are uniquely determined by the facade class, thus we can remove variation of documentation quality by developers.

The actual usage of new attributes is shown in the `test.test_autodocs.py` along with the unittest (This is good place to start review).
-- old --

### Alternative approaches

-  Option1: Copy-paste all option descriptions from the base class. 
   - Pros: Very simple, no logic is needed. User can find all available field in one place.
   - Cons: Everything is manual. If class is deeply nested, it is quite tough to manage those option docstring, and it is also quite tough to update the option docstring of the base class (we need to manually propagate the change to all subclasses). The inconsistency of docstring between subclasses are not detected by linter nor docs command.

- Option2: Just add reference to base class, i.e. See :class:\`qiskit_experiment.MyBaseClass\` for other options.
    - Pros: We don't need to manage option descriptions in subclass.
    - Cons: User cannot find whole option fields until the user reads the codebase. Usually this approach is recommended, however, in our case the method signature is just `**fields` and no option name is listed there. Thus, this approach drastically hurts usability.

Considering cons of 1&2, both approaches are not realistic, in terms of our bandwidth for code maintenance and usability. 

### Examples

This is an example of automatically generated docstring. Seems like acceptable quality.

RB Experiment documentation
![image](https://user-images.githubusercontent.com/39517270/123208978-ee4f0280-d4fa-11eb-9793-bcf4edd09270.png)

Set experiment options method
![image](https://user-images.githubusercontent.com/39517270/123209024-ff980f00-d4fa-11eb-85eb-986507490d45.png)

Set analysis options method
![image](https://user-images.githubusercontent.com/39517270/123209087-150d3900-d4fb-11eb-9f0f-5ca90c604dcd.png)

Analysis documentation
![image](https://user-images.githubusercontent.com/39517270/123209206-3ff78d00-d4fb-11eb-859a-52c39cca78d1.png)

These are automatically generated documentations.

eggerdj commented on qiskit_experiments/characterization/qubit_spectroscopy.py:
```diff
@@ -31,44 +30,57 @@
+    """Spectroscopy analysis."""
+
+    __doc_overview__ = """
+This analysis uses Gaussian function to find a peak.
```
```suggestion
This analysis uses a Gaussian function to find a peak.
```

eggerdj commented on qiskit_experiments/characterization/qubit_spectroscopy.py:
```diff
@@ -31,44 +30,57 @@
+    __doc_overview__ = """
+This analysis uses Gaussian function to find a peak.
+
+Note that this analysis assumes only single peak.
+If multiple peaks exist, you'll get a poor chi squared value.
```
```suggestion
Note that this analysis assumes only a single peak.
If multiple peaks exist, you will get a poor reduced chi squared value.
```

eggerdj commented on qiskit_experiments/characterization/qubit_spectroscopy.py:
```diff
@@ -31,44 +30,57 @@
+            name="a",
+            description="Base line.",
+            initial_guess="The maximum signal value with removed baseline.",
+            bounds="[-2, 2] scaled with maximum signal value.",
```
```suggestion
            bounds="[-2, 2] scaled to the maximum absolute signal value.",
```

eggerdj commented on qiskit_experiments/characterization/qubit_spectroscopy.py:
```diff
@@ -31,44 +30,57 @@
+            name="b",
+            description="Peak height.",
+            initial_guess="A median value of the signal.",
+            bounds="[-1, 1] scaled with maximum signal value.",
```
```suggestion
            bounds="[-1, 1] scaled to the maximum absolute signal value.",
```

eggerdj commented on qiskit_experiments/characterization/qubit_spectroscopy.py:
```diff
@@ -31,44 +30,57 @@
+        ),
+        CurveFitParameter(
+            name="f",
+            description="Center frequency. This is the fit parameter of main interest.",
```
```suggestion
            description="Center frequency. This is the fit parameter of interest.",
```

eggerdj commented on qiskit_experiments/characterization/qubit_spectroscopy.py:
```diff
@@ -31,44 +30,57 @@
+        CurveFitParameter(
+            name="f",
+            description="Center frequency. This is the fit parameter of main interest.",
+            initial_guess="A frequency value at the peak (maximum signal).",
```
```suggestion
            initial_guess="The frequency value for which the absolute signal is maximum.",
```

eggerdj commented on qiskit_experiments/characterization/qubit_spectroscopy.py:
```diff
@@ -31,44 +30,57 @@
+            description="Standard deviation of Gaussian function.",
+            initial_guess=r"Calculated from FWHM of peak :math:`w` such that "
+            r":math:`w / \sqrt{8} \ln{2}`.",
+            bounds=r"[0, :math:`\Delta x`] where :math:`\Delta x` represents frequency scan range.",
```
```suggestion
            bounds=r"[0, :math:`\Delta x`] where :math:`\Delta x` represents the scanned frequency range.",
```

eggerdj commented on qiskit_experiments/characterization/qubit_spectroscopy.py:
```diff
@@ -208,14 +228,37 @@ def _default_run_options(cls) -> Options:
+            "amp": OptionsField(
+                default=0.1,
+                annotation=float,
+                description="Amplitude of spectroscopy pulse. Usually weak power pulse is used to "
+                "suppress broadening of observed peaks.",
```
```suggestion
                description="The amplitude of the spectroscopy pulse. Usually a pulse with a weak power is used to "
                "suppress any broadening in the observed peaks.",
```

eggerdj commented on qiskit_experiments/characterization/qubit_spectroscopy.py:
```diff
@@ -208,14 +228,37 @@ def _default_run_options(cls) -> Options:
+            "duration": OptionsField(
+                default=1024,
+                annotation=int,
+                description="Duration of spectroscopy pulse. This may need to satisfy the "
+                "hardware waveform memory constraint. "
```
```suggestion
                description="The duration of the spectroscopy pulse which must satisfy any "
                "hardware waveform memory constraints. "
```

eggerdj commented on qiskit_experiments/characterization/qubit_spectroscopy.py:
```diff
@@ -208,14 +228,37 @@ def _default_run_options(cls) -> Options:
+            "sigma": OptionsField(
+                default=256,
+                annotation=Union[int, float],
+                description="Sigma of Gaussian rising and falling edges. This value should be "
+                "sufficiently smaller than the duration, "
+                "otherwise waveform is distorted. "
```
```suggestion
                description="Sigma of Gaussian rising and falling edges."
```
We do not need to enforce/suggest this, you can also run spectroscopy with a long pure Gaussian pulse and get good results.

eggerdj commented on qiskit_experiments/characterization/qubit_spectroscopy.py:
```diff
@@ -224,14 +267,7 @@ def __init__(
-        - duration: The duration of the spectroscopy pulse in samples, the default is 1000 samples.
-        - sigma: The standard deviation of the pulse, the default is duration / 4.
-        - width: The width of the flat-top in the pulse, the default is 0, i.e. a Gaussian.
+        """Create new experiment.
```
```suggestion
        """Create a new experiment.
```

eggerdj reviewed (changes requested):
This is some neat functionality. I did not review the RB experiments but focused on Spec.

nkanazawa1989 commented on qiskit_experiments/autodocs/__init__.py:
```diff
@@ -0,0 +1,44 @@
+while providing users with the standardized and high quality documentation.
+"""
+
+from . import analysis_docs, experiment_docs, set_option_docs
+from .descriptions import OptionsField, Reference, CurveFitParameter, to_options
```
@chriseclectic suggested to use relative import in another PR. I prefer absolute one but we don't have standard.

yaelbh replied:
I also prefer absolute

nkanazawa1989 commented on qiskit_experiments/autodocs/analysis_docs.py:
```diff
@@ -0,0 +1,158 @@
+        return writer.docstring
+
+
+class CurveAnalysisDocstring(_DocstringMaker):
```
I understand your point, but we cannot make this subclass. Because order of calling writer method is important to standardize docstring section ordering. Although we can make standard one a superclass of this, we need to write full `make_docstring` method for this and there is almost no benefit from doing so.

nkanazawa1989 commented on qiskit_experiments/autodocs/writer.py:
```diff
@@ -0,0 +1,260 @@
+    @abstractmethod
+    def make_docstring(cls, *args, **kwargs) -> str:
+        """Write a docstring."""
+        pass
```
No really. But I see some other abstract methods have pass (maybe personal preference).

nkanazawa1989 commented:
Thanks @eggerdj for reviewing and suggestion for wording!

eliarbel commented:
@nkanazawa1989 I think that the intention of standardizing and facilitation documentation in qiskit-experiments is great and the concrete problems you are addressing here are important. One aspect though which I think should be improved is the use of the `__doc_*__` attributes. Personally I prefer to see internal documentation comments in a certain way in the IDE, in a way which is visually consistent and differentiated from the actual code itself. With the `__doc_*__ ` attribute approach internal documentation now looks much more like the code itself rather than regular comments. Most IDEs support collapsing comments also (which I use frequently), but with this approach, developers lose that flexibility. So overall the `__doc_*__` approach seems quite disruptive from code organization and readability perspective to me. 

So I'm wondering is there is a more seamless way for supporting the same extra functionality you are introducing here, i.e without loosing the benefit of "standard" internal documentation practices. I've googled a bit to look whether there is a way to customize rst markup doc and bumped into [this](https://docutils.sourceforge.io/docs/ref/rst/directives.html#custom-interpreted-text-roles); I admit though I haven't dug into it really, it's just to show a possible approach for mitigating the concern I'm raising here. I'll be happy to discuss further.

nkanazawa1989 commented:
Thanks Eli, this is what I wanted to hear from you.

> Most IDEs support collapsing comments also (which I use frequently), but with this approach, developers lose that flexibility.

This is the point I was missing. role directives look interesting. I'll also read this carefully tomorrow!

coruscating commented:
@nkanazawa1989 I tried it out and it looks great! Some thoughts:

- Where are we putting instructions for the syntax of using this, is it staying in `test_autodocs.py`? If so, can we link that somewhere in the docs?
- Experiment options: If there’s no documentation, can it list the the parameters in returned `Options()` instead of just saying that the documentation is not provided?
- Minor issue: `typing.Union[List[float], np.array]` is rendered like
![image](https://user-images.githubusercontent.com/3870315/125156148-fed5be80-e131-11eb-87fa-a988cafe9422.png)

nkanazawa1989 commented:
Based on @eliarbel 's comment and suggestion, I modified implementation of auto documentation. 

### Previous implementation
All documentation sources are stored as a class attribute and we automatically generate `__doc__` attribute with custom class decorator when the class is loaded. The drawback of this approach is, as Eli said, we cannot collapse non-code block with IDE. This may annoy the developers.

### New implementation
Documentation is automatically formatted with [Sphinx autodoc extension](https://www.sphinx-doc.org/en/master/extdev/index.html). I added two directives `autoexperiment` and `autoanalysis`, which are run during `autosummary`,  to format class documentation by taking custom template. Basically developer need to tell template location to `autosummary` command, for example

https://github.com/Qiskit/qiskit-experiments/blob/fa8a4c0cdcc13bf51d343949692b681d50a2cf9e/qiskit_experiments/documentation/example/__init__.py#L30-L36

In this implementation, documentation is simply in the class `__doc__` attribute, with some custom section header and helper directives. You can find example in `qiskit_experiments/documentation/example` (Thanks @coruscating for suggesting to write an example!).

The downside of this approach is that user cannot get formatted document with `help` command because it is not generated until sphinx build is called. However, this will drastically reduce overhead of class import, comparing with the previous implementation. I think we can go with this approach for the first release, and if we get some user feedback regarding `help` command we can update the code base in the next release.

nkanazawa1989 commented:
This is an output example (randomly generated) of autodoc extension. Note that no module will be updated with this PR.

### Experiment class

![image](https://user-images.githubusercontent.com/39517270/125764438-379ed1cb-6f57-45d5-9a09-694c2d565eef.png)

### Analysis class

![image](https://user-images.githubusercontent.com/39517270/125764611-875da998-38ba-439e-bd34-afb4df949de5.png)

Seems like this approach can keep the quality of documentation.

eliarbel commented:
Hi Naoki, this looks very promising, thanks for addressing it! 
I've tried to use it, but hit some issue. I've added the `:template: autosummary/experiment.rst`directive to `__init__.py` of randomized_benchmarking, and running `tox -edocs`. Then I got a Sphinx error, coming from `qiskit_experiments/documentation/utils.py`. Here the last lines of the Sphinx log: 

```
    super().__init__(target_cls, docstring_lines, config, indent)
  File "/tmp/qiskit-experiments/qiskit_experiments/documentation/styles.py", line 63, in __init__
    self._parsed_lines = self._classify(lines)
  File "/tmp/qiskit-experiments/qiskit_experiments/documentation/styles.py", line 98, in _classify
    for section, lines in self._extra_sections().items():
  File "/tmp/qiskit-experiments/qiskit_experiments/documentation/styles.py", line 180, in _extra_sections
    exp_option = _generate_options_documentation(
  File "/tmp/qiskit-experiments/qiskit_experiments/documentation/utils.py", line 100, in _generate_options_documentation
    parsed_lines, target_args = _parse_option_field(
  File "/tmp/qiskit-experiments/qiskit_experiments/documentation/utils.py", line 50, in _parse_option_field
    docstring=prepare_docstring(docstring, tabsize=len(indent)), config=config
  File "/tmp/qiskit-experiments/.tox/docs/lib/python3.9/site-packages/sphinx/util/docstrings.py", line 75, in prepare_docstring
    lines = s.expandtabs(tabsize).splitlines()
AttributeError: 'NoneType' object has no attribute 'expandtabs'
```

It seems that `docstring` in `prepare_docstring` call is `None`
``` python
def _parse_option_field(
    docstring: str,
    config: SphinxConfig,
    target_args: List[str],
    indent: str = "",
) -> Tuple[List[str], List[str]]:
    """A helper function to extract descriptions of target arguments."""

    # use GoogleDocstring parameter parser
    experiment_option_parser = GoogleDocstring(
        docstring=prepare_docstring(docstring, tabsize=len(indent)), config=config
    )
``` 
Please advise

eliarbel reviewed (commented):
Do we really need to have the `documentation` directory at the top level (under `qiskit_experiments`) ? It seems like an internal util module which the user is not expected to look at or modify, SO I would expect to have its files under `docs/_ext`

nkanazawa1989 commented:
I tried that but that doesn't work. Actually it cannot port other modules from extension, perhaps some namespace issue. I'm not super familiar with the implementation of Sphinx. I'll investigate.

nkanazawa1989 commented:
Regarding the error, we can sometime solve it by removing `docs/_build` and `docs/stubs` directory before running build.

wshanks commented:
I really like the html output this produces. I think we should keep thinking about ways to improve the IDE/notebook experience. For example, this PR encourages documenting experiment options in the `_default_experiment_options` docstring, but the user will set options with `set_experiment_options` for which they will just get the generic docstring with signature `**fields`. It would be nice if they could see the actual options there. I say we should "keep thinking about it" because I don't have a good suggestion for enabling that right now. Perhaps we could make a new issue for IDE/notebook documentation improvements.

nkanazawa1989 commented:
Yes, thanks for the feedback @wshanks ! I agree IDE experience should be improved in feature. We can take two approaches AFAIK

1. Just override the method docstring with class decorator. This is easy. But user need to read documentation.

2. If we need to override function signature, we can dynamically generate method with `eval` function, i.e. generate code string and generate callable with evaluating it. I have some concern to use `eval` function.

I think we need to continue the discussion. I'll write an issue to track this issue.

nkanazawa1989 commented:
@eliarbel I found an issue. Seems like Sphinx doesn't support relative module import from extensions. Now the packages is fully under `docs/`. I'm not sure if we can import module from the protected package. See https://github.com/Qiskit/qiskit-experiments/pull/116/commits/bf9e3a4910aa4e53bd59c51470baf135dd2f5e9e Let me know if you have any suggestion.

nkanazawa1989 commented:
The sphinx build error was due to missing documentation (and error) in `CurveAnalysis` base class. Perhaps I made the mistake during resolving conflict. Fixed by https://github.com/Qiskit/qiskit-experiments/pull/116/commits/1e4dd09be722e4331215cd7fbd4486b308392062

With above fix, we can generate html documentation for RB as shown in [here](https://github.com/nkanazawa1989/qiskit-experiments/pull/1/files). This is just MVP and you need to properly update documentation with custom section and directives like [this](https://github.com/nkanazawa1989/qiskit-experiments/blob/1e4dd09be722e4331215cd7fbd4486b308392062/docs/_ext/custom_styles/example/example_experiment.py#L119-L176). I think we can do this in follow-up PR. 

Let's do dockathon! :)

eliarbel reviewed (changes requested):
From usage perspective this looks good. 

It would be good to allow seamless transition between existing docstring to the auto docstring generation format. To this end I recommend: 
1. Use the same indentation rules, i.e. don't force un-indenting actual text (e.g. the text under Overview)
2. Use the same new line rules, i.e. don't force adding new lines between the `# section: <>` lines and the actual text

nkanazawa1989 commented on docs/_ext/custom_styles/example/__init__.py:
```diff
@@ -0,0 +1,46 @@
+===========
+.. autosummary::
+    :toctree: ../stubs/
+
```
Oh thanks for finding this.

nkanazawa1989 commented:
@eliarbel Thank you so much for usability testing. Happy to hear your positive feedback.

> Use the same indentation rules, i.e. don't force un-indenting actual text (e.g. the text under Overview)

Sure. Now we _can_ use indented text block. The parser can still understand where is the section document even if indent is missing. This allows you to write document section in the exactly the same format.

> Use the same new line rules, i.e. don't force adding new lines between the # section: <> lines and the actual text

Actually it doesn't force adding the line feed (in contrast to Sphinx that forces us to remove empty line after section statement -- without error message...). You can see updated example package. I just removed empty line (now it really looks like previous format) and it still works.

These changes are made in this commit https://github.com/Qiskit/qiskit-experiments/pull/116/commits/f6cda8f81fd44a21aaf89c7ab0b82f3515c5934f

You can also check updated RB example [here](https://github.com/nkanazawa1989/qiskit-experiments/pull/1/files) -- please ignore two files in the beginning.

eliarbel reviewed (approved):
I think this is ready for production!

coruscating commented:
@eggerdj this is pending your approval, can you give a look?
